hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/EncryptionS3ClientFactory.java

```diff
@@ -0,0 +1,299 @@
+      throws IOException {
+    if (!isEncryptionClientAvailable()) {
+      throw unavailable(uri, ENCRYPTION_CLIENT_CLASSNAME, null,
+          "No encryption client available");
+    }
+
+    s3Client = super.createS3Client(uri, parameters);
+    s3AsyncClient = super.createS3AsyncClient(uri, parameters);
+
+    return createS3EncryptionClient(parameters);
+  }
+
+  /**
+   * Create async encrypted s3 client.
+   * @param uri S3A file system URI
+   * @param parameters parameter object
+   * @return async encrypted s3 client
+   * @throws IOException IO failures
+   */
+  @Override
+  public S3AsyncClient createS3AsyncClient(URI uri, S3ClientCreationParameters parameters)
+      throws IOException {
+    if (!isEncryptionClientAvailable()) {
+      throw unavailable(uri, ENCRYPTION_CLIENT_CLASSNAME, null,
+          "No encryption client available");
+    }
+    return createS3AsyncEncryptionClient(parameters);
+  }
+
+  /**
+   * Creates an S3EncryptionClient instance based on the provided parameters.
+   *
+   * @param parameters The S3ClientCreationParameters containing the necessary configuration.
+   * @return An instance of S3EncryptionClient.
+   */
+  private S3Client createS3EncryptionClient(S3ClientCreationParameters parameters) {
+    CSEMaterials cseMaterials = parameters.getClientSideEncryptionMaterials();
+    Preconditions.checkArgument(s3AsyncClient != null,
+        "S3 async client not initialized");
+    Preconditions.checkArgument(s3Client != null,
+        "S3 client not initialized");
+    Preconditions.checkArgument(parameters != null,
+        "S3ClientCreationParameters is not initialized");
+
+    S3EncryptionClient.Builder s3EncryptionClientBuilder =
+        S3EncryptionClient.builder()
+            .wrappedAsyncClient(s3AsyncClient)
+            .wrappedClient(s3Client)
+            // this is required for doing S3 ranged GET calls
+            .enableLegacyUnauthenticatedModes(true)
+            // this is required for backward compatibility with older encryption clients
+            .enableLegacyWrappingAlgorithms(true);
+
+    switch (cseMaterials.getCseKeyType()) {
+    case KMS:
+      Keyring kmsKeyring = createKmsKeyring(parameters, cseMaterials);
+      CryptographicMaterialsManager kmsCryptoMaterialsManager =
+          DefaultCryptoMaterialsManager.builder()
+              .keyring(kmsKeyring)
+              .build();
+      s3EncryptionClientBuilder.cryptoMaterialsManager(kmsCryptoMaterialsManager);
+      break;
+    case CUSTOM:
+      Keyring keyring = getKeyringProvider(cseMaterials.getCustomKeyringClassName(),
+          cseMaterials.getConf());
+      CryptographicMaterialsManager customCryptoMaterialsManager =
+          DefaultCryptoMaterialsManager.builder()
+              .keyring(keyring)
+              .build();
+      s3EncryptionClientBuilder.cryptoMaterialsManager(customCryptoMaterialsManager);
+      break;
+    default:
+      break;
```

Comment:
There will be always a match because, this method is called after `getClientSideEncryptionMaterials` which constructs the CSEMaterial object and it throws exception if KMS or CUSTOM is not the CSE algorithm.